libyul/Object.cpp

```diff
@@ -134,12 +134,20 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
 		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
 		yulAssert(
 			subIndexIt != object->subIndexByName.end(),
-			"Assembly object <" + _qualifiedName.str() + "> not found or does not contain code."
+			"Assembly object <" + _qualifiedName.str() + "> not found."
 		);
-		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
-		yulAssert(object, "Assembly object <" + _qualifiedName.str() + "> not found or does not contain code.");
-		yulAssert(object->subId != numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+		auto const* objectNode = object->subObjects[subIndexIt->second].get();
+		object = dynamic_cast<Object const*>(objectNode);
+		if (object)
+		{
+			yulAssert(object->subId != numeric_limits<size_t>::max());
+			path.push_back({object->subId});
+		}
+		else
+			yulAssert(
```

Comment:
This should at least assert that we are in the last component of the path somehow, or that `object` is not `nullptr` at the beginning of the next iteration.

Reply:
> This should at least assert that we are in the last component of the path somehow, or that `object` is not `nullptr` at the beginning of the next iteration.

@chriseth  sure I will add it, but please give an advice of how such expression might look like. If write something like

`ObjectA.DataObjectA.ObjectB` it causes an error during the parsing:

```
Error parsing source.
Error: Unknown data object "ObjectA.DataObjectA.ObjectB".
 --> source:7:22:
  |
7 |         pop(datasize("ObjectA.DataObjectA.ObjectB"))
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
If it somehow can happen I think it makes sense to add an appropriate test.

Reply:
It can't happen, which is why it should be an assert in the code, not a test. But the only reason it can't happen is that [we have a validation against it elsewhere](https://github.com/ethereum/solidity/blob/10f81425a375d4d2622e8aa22ec7a9fb6dec30e6/libyul/AsmAnalysis.cpp#L393-L401). If you called the function directly with `ObjectA.DataObjectA.ObjectB`, it would crash because it continues the loop and tries to use `->` on a null `object`. There should be an assert at the beginning of the loop checking that `object` is not null, as @chriseth suggested, so that if anyone in the future tries to use that function without realizing that a separate validation is needed, the assert will fail instead of letting the function access memory via a null pointer.

But actually, I see we don't have a test against `ObjectA.DataObjectA.ObjectB` in `test/libyul/yulSyntaxTests/objects/` so please add one too. But that's to test the validation, not the function. The assert is needed regardless of this.

Would be a good idea to check if all important corner cases we could run into with nested data that are invalid are covered - and if not, add tests for them.

Another assert checking that we're at the last component might be useful too, but that's more tricky. We'd have to add an index to the loop or compare the length of the resulting `path` against `subObjectPathComponents`. I think that checking if `object` is not null is enough and will cover the same situations anyway.

Reply:
@cameel thank you very much for the detailed explanation, appreciated. This is what I looked for. Will update this PR.